client/shared/src/search/parser/treeParser.ts

```diff
@@ -0,0 +1,187 @@
+export interface Leaf {
+    type: 'leaf'
+    value: string
+}
+
+enum OperatorKind {
+    Or = 'or',
+    And = 'and',
+}
+
+/**
+ * A nonterminal node for operators 'and' and 'or'.
+ */
+export interface Operator {
+    type: 'operator'
+    operands: Node[]
+    kind: OperatorKind
+}
+
+export type Node = Leaf | Operator
+
+interface ParseError {
+    type: 'error'
+    expected: string
+}
+
+export interface ParseSuccess {
+    type: 'success'
+    nodes: Node[]
+}
+
+export type ParseResult = ParseError | ParseSuccess
+
+export const toString = (node: Node): string => {
+    const result: string[] = []
+    switch (node.type) {
+        case 'operator':
+            for (const operand of node.operands) {
+                result.push(toString(operand))
+            }
+            switch (node.kind) {
+                case OperatorKind.Or:
+                    return `(or ${result.join(' ')})`
+                case OperatorKind.And:
+                    return `(and ${result.join(' ')})`
+            }
+        case 'leaf':
+            return node.value
+    }
+}
+
+interface State {
+    result: ParseResult
+    advance: number
+}
+
+const match = (input: string, value: string): boolean => input.startsWith(value)
+
+const newNodes = (nodes: Node[]): ParseResult => ({
+    type: 'success',
+    nodes,
+})
+
+const newOperator = (nodes: Node[], kind: OperatorKind): ParseResult => ({
+    type: 'success',
+    nodes: [
+        {
+            type: 'operator',
+            operands: nodes,
+            kind,
+        },
+    ],
+})
+
+const newLeaf = (value: string): ParseResult => ({
+    type: 'success',
+    nodes: [
+        {
+            type: 'leaf',
+            value,
+        },
+    ],
+})
+
+/**
+ * parses tokens that are leaves in the tree. {@link State} returns the result of the parse, and how much the caller should advance input.
+ */
+export const scanLeaf = (input: string): State => {
+    let position = 0
+    let current = ''
+    const result: string[] = []
+    while (input.length > 0) {
+        current = input[0]
+        position++
+        input = input.slice(1)
+        if (current === ' ') {
+            position-- // Backtrack.
+            break
+        }
+        result.push(current)
+    }
+    return { result: newLeaf(result.join('')), advance: position }
```

Comment:
I think this is strictly equivalent (but I guess this function is going to do a bit more work once we start introducing parens, so the simplification might not be worth it here):

```suggestion
    const i = input.indexOf(' ')
    if (i === -1) {
        return { result: newLeaf(input), advance: input.length }
    }
    return { result: newLeaf(input.slice(0, i)), advance: i }
```